src/js/node/events.js

```diff
@@ -356,52 +357,67 @@ function once(emitter, type, options) {
   });
 }
 
-async function* on(emitter, event, options = {}) {
+function on(emitter, event, options = {}) {
   const signal = options.signal;
-  if (signal?.aborted) throw new AbortError(undefined, { cause: signal?.reason });
 
   const { FixedQueue } = require("internal/fixed_queue");
   const unconsumedPromises = new FixedQueue();
   const unconsumedEvents = new FixedQueue();
   const unconsumedErrors = new FixedQueue();
   let done = false;
 
-  emitter.on(event, ev => {
+  const eventHandlerBody = ev => {
+    // If there is a pending Promise -> resolve with current event value.
     if (!unconsumedPromises.isEmpty()) {
       const { resolve } = unconsumedPromises.shift();
-      return resolve([ev]);
+      return resolve(ev);
     }
-    unconsumedEvents.push([ev]);
-  });
-  emitter.on("error", ex => {
+    // Else: Add event value to queue so it can be consumed by a future Promise.
+    unconsumedEvents.push(ev);
+  };
+  const eventHandler = options[kFirstEventParam] ? eventHandlerBody : (...args) => eventHandlerBody(args)
+  emitter.on(event, eventHandler);
+
+  const errorHandler = ex => {
     if (!unconsumedPromises.isEmpty()) {
       const { reject } = unconsumedPromises.shift();
       return reject(ex);
     }
     unconsumedErrors.push(ex);
-  });
+  }
+  emitter.on("error", errorHandler);
+
   signal?.addEventListener("abort", () => {
     emitter.emit("error", new AbortError(undefined, { cause: signal?.reason }));
   });
 
+  // If any of the close events is emitted -> remove listeners
+  // and yield only the remaining queued-up values in iterator.
   for (const evName of options?.close || []) {
     emitter.on(evName, () => {
-      emitter.emit(event, undefined);
+      emitter.removeListener(event, eventHandler)
+      emitter.removeListener("error", errorHandler)
```

Comment:
Instead of checking for `done` in the listener body, I think it's cleaner to just remove the listeners.